packages/hardhat/contracts/LargeGrant.sol

```diff
@@ -0,0 +1,338 @@
+			uint256 amount = _milestoneAmounts[i];
+			if (amount == 0) {
+				revert MilestoneZeroAmount();
+			}
+			stage.milestones.push(
+				Milestone({
+					number: uint8(i + 1),
+					amount: amount,
+					completed: false
+				})
+			);
+		}
+
+		emit AddGrantStage(
+			_grantId,
+			uint8(grant.stages.length),
+			_milestoneAmounts
+		);
+	}
+
+	function completeMilestone(
+		uint256 _grantId,
+		uint8 _stageNumber,
+		uint8 _milestoneNumber
+	) public whenNotPaused onlyRole(OWNER_ROLE) {
+		GrantData storage grant = grants[_grantId];
+
+		if (grant.builder == address(0)) {
+			revert GrantDoesNotExist();
+		}
+
+		if (_stageNumber > grant.stages.length) {
+			revert WrongStageNumber();
+		}
+
+		if (
+			_milestoneNumber > grant.stages[_stageNumber - 1].milestones.length
+		) {
+			revert WrongMilestoneNumber();
+		}
+
+		if (
+			grant
+				.stages[_stageNumber - 1]
+				.milestones[_milestoneNumber - 1]
+				.completed
+		) {
+			revert MilestoneAlreadyCompleted();
+		}
+
+		uint256 amount = grant
+			.stages[_stageNumber - 1]
+			.milestones[_milestoneNumber - 1]
+			.amount;
+
+		if (ERC20(tokenAddress).balanceOf(address(this)) < amount)
+			revert InsufficientContractFunds();
+
+		ERC20(tokenAddress).transfer(grant.builder, amount);
+
+		grant
+			.stages[_stageNumber - 1]
+			.milestones[_milestoneNumber - 1]
+			.completed = true;
+
+		emit CompleteMilestone(
+			_grantId,
+			_stageNumber,
+			_milestoneNumber,
+			amount
+		);
+	}
+
+	function completeMilestones(
+		uint256 _grantId,
+		uint8 _stageNumber,
+		uint8[] memory _milestoneNumbers
+	) public whenNotPaused onlyRole(OWNER_ROLE) {
+		GrantData storage grant = grants[_grantId];
+
+		if (grant.builder == address(0)) {
+			revert GrantDoesNotExist();
+		}
+
+		if (_stageNumber > grant.stages.length) {
+			revert WrongStageNumber();
+		}
+
+		uint256 totalAmount = 0;
+
+		for (uint i = 0; i < _milestoneNumbers.length; i++) {
+			uint8 _milestoneNumber = _milestoneNumbers[i];
+
+			if (
+				_milestoneNumber >
+				grant.stages[_stageNumber - 1].milestones.length
+			) {
+				revert WrongMilestoneNumber();
+			}
+
+			if (
+				grant
+					.stages[_stageNumber - 1]
+					.milestones[_milestoneNumber - 1]
+					.completed
+			) {
+				revert MilestoneAlreadyCompleted();
+			}
+
+			uint256 amount = grant
+				.stages[_stageNumber - 1]
+				.milestones[_milestoneNumber - 1]
+				.amount;
+
+			totalAmount += amount;
+
+			grant
+				.stages[_stageNumber - 1]
+				.milestones[_milestoneNumber - 1]
+				.completed = true;
+
+			emit CompleteMilestone(
+				_grantId,
+				_stageNumber,
+				_milestoneNumber,
+				amount
+			);
+		}
+
+		if (ERC20(tokenAddress).balanceOf(address(this)) < totalAmount)
+			revert InsufficientContractFunds();
+
+		ERC20(tokenAddress).transfer(grant.builder, totalAmount);
+	}
+
+	function addOwner(address newOwner) public onlyRole(DEFAULT_ADMIN_ROLE) {
+		grantRole(OWNER_ROLE, newOwner);
+		emit AddOwner(newOwner, msg.sender);
```

Comment:
I implemented this like it was in the Stream contract, but the AddOwner is unnecessary. The AccessControl extension will emit a RoleGranted (or RoleRevoked) event anyway. Maybe there was a reason to emit this new event cc/ @Pabl0cks @technophile-04 @rin-st

Reply:
Ohh yeah we could remove it 🙌  We just missed that in `Stream contract` and also we are not using that anywhere too